File: 3_Structural_Design_Patterns/3.02_Decorator/Python_Snippet/Decorator.py

class Beverage:
    def __init__(self, description):
        self.description = description

    def get_description(self):
        return self.description

    def cost(self):
        return 0.0

class Espresso(Beverage):
    def __init__(self):
        super().__init__("Espresso")

    def cost(self):
        return 1.99

class Decorator(Beverage):
    def __init__(self, beverage):
        self.beverage = beverage

    def get_description(self):
        return self.beverage.get_description() + " " + self.description

    def cost(self):
        return self.beverage.cost() + self.additional_cost

class Milk(Decorator):
    def __init__(self, beverage):
        super().__init__(beverage)
        self.description = "Milk"
        self.additional_cost = 0.50

class Sugar(Decorator):
    def __init__(self, beverage):
        super().__init__(beverage)
        self.description = "Sugar"
        self.additional_cost = 0.25

class WhippedCream(Decorator):
    def __init__(self, beverage):
        super().__init__(beverage)
        self.description = "Whipped Cream"
        self.additional_cost = 1.00

File: 3_Structural_Design_Patterns/3.04_Decorator/Java_Snippet/Decorator.java
interface Beverage {
    String getDescription();

    double getCost();
}

class Espresso implements Beverage {
    @Override
    public String getDescription() {
        return "Espresso";
    }

    @Override
    public double getCost() {
        return 1.99;
    }
}

abstract class Decorator implements Beverage {
    private final Beverage beverage;

    abstract String getDecoratorDescription();

    abstract double getAdditionalCost();

    public Decorator(Beverage beverage) {
        this.beverage = beverage;
    }

    @Override
    public String getDescription() {
        return beverage.getDescription() + " " + getDecoratorDescription();
    }

    @Override
    public double getCost() {
        return beverage.getCost() + getAdditionalCost();
    }
}

class Milk extends Decorator {

    public Milk(Beverage beverage) {
        super(beverage);
    }

    @Override
    String getDecoratorDescription() {
        return "Milk";
    }

    @Override
    double getAdditionalCost() {
        return 0.50;
    }

}

class Sugar extends Decorator {
    public Sugar(Beverage beverage) {
        super(beverage);
    }

    @Override
    String getDecoratorDescription() {
        return "Sugar";
    }

    @Override
    double getAdditionalCost() {
        return 0.25;
    }
}

class WhippedCream extends Decorator {
    public WhippedCream(Beverage beverage) {
        super(beverage);
    }

    @Override
    String getDecoratorDescription() {
        return "Whipped Cream";
    }

    @Override
    double getAdditionalCost() {
        return 1.00;
    }
}